pkg/io/v2/providers/providers.go

```diff
@@ -0,0 +1,235 @@
+
+func GetBucket(ctx context.Context, credentials []byte, path string) (*blob.Bucket, error) {
+	storageProvider, bucket, _, err := ParseStoragePath(path)
+	if err != nil {
+		return nil, err
+	}
+
+	switch storageProvider {
+	case providerFile:
+		return getFileBucket(bucket)
+	case providerGS:
+		return getGCSBucket(ctx, credentials, bucket)
+	case providerS3:
+		return getS3Bucket(ctx, credentials, bucket)
+	default:
+		return nil, fmt.Errorf("unknown storageProvider: %s", storageProvider)
+	}
+}
+
+func getFileBucket(bucket string) (*blob.Bucket, error) {
+	bkt, err := blob.OpenBucket(context.Background(), "file://"+bucket)
+	if err != nil {
+		return nil, fmt.Errorf("error opening file bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+func getGCSBucket(ctx context.Context, credentials []byte, bucketName string) (*blob.Bucket, error) {
+	googleCredentials, err := google.CredentialsFromJSON(ctx, credentials, storage.ScopeFullControl)
+	if err != nil {
+		return nil, fmt.Errorf("error getting Google credentials from JSON: %v", err)
+	}
+
+	client, err := gcp.NewHTTPClient(
+		gcp.DefaultTransport(),
+		gcp.TokenSource(googleCredentials.TokenSource))
+	if err != nil {
+		return nil, fmt.Errorf("error creating GCP Http Client: %v", err)
+	}
+
+	bkt, err := gcsblob.OpenBucket(ctx, client, bucketName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error opening GCS bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+type s3Credentials struct {
+	Region           string `json:"region"`
+	Bucket           string `json:"bucket"`
+	Endpoint         string `json:"endpoint"`
+	Insecure         bool   `json:"insecure"`
+	S3ForcePathStyle bool   `json:"s3_force_path_style"`
+	AccessKey        string `json:"access_key"`
+	SecretKey        string `json:"secret_key"`
+}
+
+func getS3Bucket(ctx context.Context, creds []byte, bucketName string) (*blob.Bucket, error) {
+	s3Credentials := &s3Credentials{}
+	if err := json.Unmarshal(creds, s3Credentials); err != nil {
+		return nil, fmt.Errorf("error getting S3 credentials from JSON: %v", err)
+	}
+
+	staticCredentials := credentials.NewStaticCredentials(s3Credentials.AccessKey, s3Credentials.SecretKey, "")
+
+	sess, err := session.NewSession(&aws.Config{
+		Credentials:      staticCredentials,
+		Endpoint:         aws.String(s3Credentials.Endpoint),
+		DisableSSL:       aws.Bool(s3Credentials.Insecure),
+		S3ForcePathStyle: aws.Bool(s3Credentials.S3ForcePathStyle),
+		Region:           aws.String(s3Credentials.Region),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error creating S3 Session: %v", err)
+	}
+
+	bkt, err := s3blob.OpenBucket(ctx, sess, bucketName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error opening S3 bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+func SignedURL(ctx context.Context, credentials []byte, storagePath string, opts *blob.SignedURLOptions) (string, error) {
+	storageProvider, bucketName, relativePath, err := ParseStoragePath(storagePath)
```

Comment:
This func (if it needs to exist in the first place, see other comment), could be drastically simplified by only having provider-specific handling for google:

```
if !storageProvider == providerGS {
// Do GCS anonymous specific stuff
}
// GetBucket
// HandleErr
// Get Signed URL

Reply:
Sounds good. I'll adjust this and then move the method to the next PR